assembly/src/assembler/instruction/procedures.rs

```diff
@@ -31,9 +31,19 @@ impl Assembler {
 
         // get the procedure from the assembler
         let proc_cache = self.proc_cache.borrow();
-        let proc = proc_cache.get_by_id(proc_id).expect("procedure not in cache");
-        debug_assert!(proc.is_export(), "not imported procedure");
 
+        let proc = match proc_cache.get_reexported_proc_ref_id(proc_id) {
+            Some(ref_proc_id) => {
+                // get the re-exported procedure
+                proc_cache.get_by_id(ref_proc_id).expect("procedure not in cache")
+            }
+            None => {
+                // get the imported procedure
+                proc_cache.get_by_id(proc_id).expect("procedure not in cache")
+            }
+        };
```

Comment:
In the case where there is a name clash between re-exported and locally defined procedures, this bit of code gives precedence to the re-exported procedure, which in effect means that re-exported procedures shadow locally defined ones.

I don't think we should allow shadowing, but even if we do I think we should enforce it in the parser rather than here, i.e., the AST should only contain the procedures that are actually visible to the outside world.

Reply:
Thanks, I missed that earlier. I've added a check to make sure the duplicate ids/names are not allowed.

Updated insertions:

```
    // MUTATORS
    // --------------------------------------------------------------------------------------------
    /// Inserts a [Procedure] into the [ProcedureCache].
    pub fn insert(&mut self, proc: Procedure) -> Result<(), AssemblyError> {
        // if a re-exported procedure with the same id is already in the cache, return an error
        if self.proc_aliases.contains_key(proc.id()) {
            return Err(AssemblyError::duplicate_proc_id(proc.id()));
        }
        // If the entry is `Vacant` then insert the Procedure. If the `ProcedureId` is already in
        // the cache (i.e. it is a duplicate) then return an error.
        match self.proc_map.entry(*proc.id()) {
            Entry::Occupied(_) => Err(AssemblyError::duplicate_proc_id(proc.id())),
            Entry::Vacant(entry) => {
                let mast_root = proc.code_root().hash();
                self.mast_map.entry(mast_root).or_insert(*proc.id());
                entry.insert(proc);
                Ok(())
            }
        }
    }

    /// Inserts a re-exported [ProcedureId] to reference [ProcedureId] map into the
    /// [ProcedureCache].
    pub fn insert_proc_alias(
        &mut self,
        alias_proc_id: ProcedureId,
        proc_id: ProcedureId,
    ) -> Result<(), AssemblyError> {
        // if a procedure with the same id is already in the cache, return an error
        if self.proc_map.contains_key(&alias_proc_id) {
            return Err(AssemblyError::duplicate_proc_id(&alias_proc_id));
        }
        // If the entry is `Vacant` then insert the ProcedureId of the alias procedure. If the
        // `ProcedureId` is already in the cache (i.e. it is a duplicate) then return an error.
        match self.proc_aliases.entry(alias_proc_id) {
            Entry::Occupied(_) => Err(AssemblyError::duplicate_proc_id(&alias_proc_id)),
            Entry::Vacant(entry) => {
                entry.insert(proc_id);
                Ok(())
            }
        }
    }
```

Reply:
Thanks, I've added a check during insertions in `ProcedureCache`.

In my understanding, if we go with checking for duplicates during parsing, we will have to check for duplicates separately for cases when the modules are imported from libraries (deserialization) or source files. It would be great to get @bobbinth's opinion on this as well.

Reply:
I agree with @jjcnn that shadowing should not be allowed. Regarding checking for duplicates separately, could you give a bit more info why that would be the case?

Reply:
I meant if we move the duplicate check to the parser when `ModuleAst` is built, it will be slightly inefficient and will require small amount of code duplication. And removing the redundant check during insertions in `ProcedureCache` also seems wrong for some reason even though we resolve duplicates during parsing.

To elaborate, if we read the module from a serialized masl library, we get two different vectors of `ProcedureAst` and `ProcReExport` and we need to check which elements are duplicates across the two vectors which is a little inefficient. When we read the module from a masm file, we would probably want to move this check to `parse_procedures` to do it cheaper as we have maps for both local procedures and re-exported procedures.

After we have checked for duplicates during parsing, then when we insert the procedures in `proc_map` and `proc_aliases`, we can (I think) safely avoid checking for duplicates there but I'm not sure if we'd want to remove that redundant check from there. If we instead just check for duplicates while inserting in `ProcedureCache` maps, that still catches all the cases irrespective of how the module was imported.

I don't have strong opinions on this though if the convention is to do these kinds of checks during parsing.

Reply:
This problem becomes simpler to solve if there is only a single collection for both types of procedures, right?